packages/scripts/src/upgrade-tests/voting2/3_SimulateVoting.ts

```diff
@@ -0,0 +1,687 @@
+}
+
+const zeroBigNumber: BigNumber = hre.ethers.BigNumber.from(0);
+
+// Constants hardcoded in the SlashingLibrary, needs to be updated here upon change.
+const wrongVoteSlashPerToken = parseEther("0.0016");
+const noVoteSlashPerToken = parseEther("0.0016");
+
+// Initial voter balances relative to GAT.
+const voter1RelativeGatFunding = parseEther("0.6");
+const voter2RelativeGatFunding = parseEther("0.55");
+const voter3RelativeGatFunding = parseEther("0.5");
+
+// Tested price identifier should be whitelisted.
+const priceIdentifier = formatBytes32String("NUMERICAL");
+
+async function main() {
+  const rewardTrackers: RewardTrackers = {
+    lastUpdateTimestamp: zeroBigNumber,
+    staked: [zeroBigNumber, zeroBigNumber, zeroBigNumber],
+    unclaimedRewards: [zeroBigNumber, zeroBigNumber, zeroBigNumber],
+  };
+
+  // Initiates data request through Optimistic Oracle by requesting, proposing and diputing.
+  // Returns stamped ancillary data to be used in voting.
+  async function _requestProposeDispute(priceRequestData: PriceRequestData): Promise<BytesLike> {
+    await (
+      await optimisticOracleV2
+        .connect(requesterSigner)
+        .requestPrice(
+          priceRequestData.priceRequest.identifier,
+          priceRequestData.priceRequest.time,
+          priceRequestData.originalAncillaryData,
+          votingToken.address,
+          0
+        )
+    ).wait();
+    await (await votingToken.connect(requesterSigner).approve(optimisticOracleV2.address, finalFee.mul(4))).wait();
+    await (
+      await optimisticOracleV2
+        .connect(requesterSigner)
+        .proposePrice(
+          requesterSigner.address,
+          priceRequestData.priceRequest.identifier,
+          priceRequestData.priceRequest.time,
+          priceRequestData.originalAncillaryData,
+          priceRequestData.proposedPrice
+        )
+    ).wait();
+    await (
+      await optimisticOracleV2
+        .connect(requesterSigner)
+        .disputePrice(
+          requesterSigner.address,
+          priceRequestData.priceRequest.identifier,
+          priceRequestData.priceRequest.time,
+          priceRequestData.originalAncillaryData
+        )
+    ).wait();
+    return await optimisticOracleV2.stampAncillaryData(priceRequestData.originalAncillaryData, requesterSigner.address);
+  }
+
+  // Construct voting structure for commit and reveal.
+  async function _createVote(priceRequest: VotingPriceRequest, voter: string, price: string): Promise<CommittedVote> {
+    const salt = getRandomSignedInt().toString();
+    const roundId = Number(await votingV2.getCurrentRoundId());
+    const voteHash = computeVoteHashAncillary({
+      price,
+      salt,
+      account: voter,
+      time: Number(priceRequest.time),
+      roundId,
+      identifier: priceRequest.identifier.toString(),
+      ancillaryData: priceRequest.ancillaryData.toString(),
+    });
+    return <CommittedVote>{ priceRequest, salt, price, voteHash };
```

Comment:
I've never seen this syntax before, what is this doing?

Reply:
The returned `CommittedVote` contains all the data that can be reused both when committing and revealing.

This facilitates more readable main test logic, e.g. when `_commitVote` returns `voter1FirstRequestVote` we can:

```
await _commitVote(voter1Signer, voter1FirstRequestVote);
```

and then

```
await _revealVote(voter1Signer, voter1FirstRequestVote);
```